### PR TITLE
Feature/finetune first model

### PR DESCRIPTION
**Pull request by JeroenMandersloot**
Created a dummy pipeline for finetuning the VGGFace model in `finetuning.py`.

Changes:
* Refactored deepface folder: removed all deepface repo-specific files and only kept the source files (in effect, the contents of what used to be the `/deepface/deepface` folder have been moved up one folder)
* Implemented a generic `FinetuneModel` class that can serve as a wrapper around a tensorflow model that can then easily be finetuned using a triplet loss.
* `VGGFace` updates:
  * Fixed the layer that was used to output embeddings (previous configuration was wrong I think)
  * Now uses `tf.keras` instead of `keras` so we can use both tensorflow and keras methods
  * Now outputs L2-normalized embeddings
* When adding finetuning support for the other Deepface models, the same should be done for them, but these are relatively minor changes. Afterwards, the existing pipeline should be generic enough to incorporate all of them out of the box.
* Added some unit tests for sanity checks.

### CODE REVIEW

File: deepface/basemodels/VGGFace.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(Convolution2D(4096, (7, 7), activation='relu'))
+    model.add(Dropout(0.5))
+    model.add(Convolution2D(4096, (1, 1), activation='relu'))
+    model.add(Dropout(0.5))
+    model.add(Convolution2D(2622, (1, 1)))
```
Comment (rolfypma) [inline deepface/basemodels/VGGFace.py:58]: hebben we deze laatste paar lagen dan nog wel nodig? ik snap niet helemaal wat hier gebeurt geloof ik

File: lr_face/models.py [hunk 2]
```diff
@@ -54,5 +55,96 @@ def score_for_pair(self, pair):
...
+    """
+    A subclass of tf.keras.Model that can be used to finetune the pre-trained
+    embedding models.This new training model takes 3 inputs, namely:
+
+        anchor: A 4D tensor containing a batch of anchor images with shape
+            `(batch_size, height, width, num_channels)`.
+        positive: A 4D tensor containing a batch of images of the same identity
+            as the anchor image with shape `(batch_size, height, width,
+            num_channels)`.
+        positive: A 4D tensor containing a batch of images of a different
```
Comment (rolfypma) [inline lr_face/models.py:73]: two positives! :)

File: lr_face/losses.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        # Squeeze the 2nd dimension of each of the split embedding
+        # tensors. Resulting shape: `(batch_size, embedding_size)`.
+        anchor = tf.squeeze(anchor, axis=1)
+        positive = tf.squeeze(positive, axis=1)
+        negative = tf.squeeze(negative, axis=1)
+
+        # Compute the euclidean distance between the anchor and the two
+        # query images. Resulting shape: `(batch_size)`.
+        positive_distance = tf.reduce_sum(tf.square(anchor - positive), axis=1)
```
Comment (rolfypma) [inline lr_face/losses.py:51]: klopt het dat we hier euclidean gebruiken en elders cosine distance?

File: lr_face/models.py [hunk 4]
```diff
@@ -54,5 +55,96 @@ def score_for_pair(self, pair):
...
+    def __init__(self, embedding_layer: tf.keras.Model, *args, **kwargs):
+        """
+        Arguments:
+            embedding_layer: A tf.keras.Model instance that given a batch of
+                images computes their embeddings. It should accept 4D tensors
+                with shape `(batch_size, height, width, num_channels)` as input
+                and output 2D tensors of shape `(batch_size, embedding_size)`.
+        """
+        super().__init__(*args, **kwargs)
+        self.embedding_layer = embedding_layer
```
Comment (rolfypma) [inline lr_face/models.py:93]: is embedding layer hier niet een verwarrende naam? het is een heel model dat als laatste layer een embedding layer heeft?

File: deepface/basemodels/VGGFace.py (hunk 1 above)
Comment (JeroenMandersloot) [reply deepface/basemodels/VGGFace.py:58]: Het model zoals dat nu in `baseModel()` gedefinieerd is, is een classificatiemodel. Die `2622` die je ziet, is het totaal aantal identiteiten in de dataset waar het op getraind is. Met die architectuur zijn de gewichten opgeslagen, dus om die gewichten vervolgens te laden moet je dezelfde architectuur nabouwen. Wil je het model daarna niet voor classificatie maar voor het genereren van embeddings gebruiken, dan haal je de laatste paar lagen eraf. 

In dit geval zijn de laatste 5 lagen de volgende:
```python
	model.add(Convolution2D(4096, (1, 1), activation='relu'))
	model.add(Dropout(0.5))
	model.add(Convolution2D(2622, (1, 1)))
	model.add(Flatten())
	model.add(Activation('softmax'))
```

Hiervan is de eerste eigenlijk de embedding laag. De volgende 4 lagen kun je eigenlijk zien als een enkele dense laag met een softmax activatie om vervolgens class probabilities van de 2622 classes (identities) te outputten. Vandaar dat het mij logisch leek om die `model.add(Convolution2D(4096, (1, 1), activation='relu'))` als uiteindelijk embedding laag te nemen.

File: lr_face/losses.py (hunk 3 above)
Comment (JeroenMandersloot) [reply lr_face/losses.py:51]: Klopt ja, de triplet loss is standaard gedefinieerd met euclidean distance; dus we moeten even kijken hoe we dat gelijk willen trekken.

File: lr_face/models.py (hunk 4 above)
Comment (JeroenMandersloot) [reply lr_face/models.py:93]: Ja het ligt er een beetje aan hoe je er tegen aan kijkt. In Keras terminologie is `Model` een subclass van `Layer` en daarmee elk model ook een layer, maar ik snap je punt. Ik kan het hernoemen naar `embedding_model`.

File: lr_face/models.py (hunk 2 above)
Comment (JeroenMandersloot) [reply lr_face/models.py:73]: Oeps!

File: deepface/basemodels/VGGFace.py (hunk 1 above)
Comment (rolfypma) [reply deepface/basemodels/VGGFace.py:58]: daar was ik al bang voor. dus dit repo doet het zelf 'fout'? Ben benieuwd of deze verbetering gaat helpen met performance:)

File: finetuning.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    To load the embedding model for VGGFace for example, you would use:
+
+        `BaseModel.VGGFace.load_embedding_model()`
+
+    Similarly, to load a finetune model, you would use:
+
+        `BaseModel.VGGFace.load_finetune_model()`
+    """
+    VGGFace = auto()
```
Comment (rolfypma) [inline finetuning.py:37]: is hier simpel support voor lowercase in te butsen? nu moet je precies de hoofdletters volgen (licht gedoe)

File: lr_face/models.py [hunk 6]
```diff
@@ -54,5 +55,96 @@ def score_for_pair(self, pair):
         img2 = resize_and_normalize(pair[1], self.model.input_shape[1:3])
         img1_representation = self.model.predict(img1)[0, :]
         img2_representation = self.model.predict(img2)[0, :]
-        score = spatial.distance.cosine(img1_representation, img2_representation)
+        score = spatial.distance.cosine(img1_representation,
+                                        img2_representation)
         return score
+
+
+class FinetuneModel(tf.keras.Model):
```
Comment (rolfypma) [inline lr_face/models.py:63]: dit is feitelijk een schil om je embedding model heen om m 3x aan te kunnen roepen. Hoewel ik snap dat dit een keras Model is denk ik dat een naam als 'TripletEmbedder' of iets in die trend beschrijvender is.

File: lr_face/models.py [hunk 7]
```diff
@@ -54,5 +55,96 @@ def score_for_pair(self, pair):
...
+    """
+    A subclass of tf.keras.Model that can be used to finetune the pre-trained
+    embedding models.This new training model takes 3 inputs, namely:
```
Comment (rolfypma) [inline lr_face/models.py:66]: Enigszins ambigu omdat zowel init als call argumenten nemen. Misschien toevoegen iets als. 
This class wraps a particular embedding model 'finetuner = TripletEmbedder(embedding_model)'. When called, the Embedder takes 3 inputs, namely: ...

File: deepface/basemodels/VGGFace.py (hunk 1 above)
Comment (JeroenMandersloot) [reply deepface/basemodels/VGGFace.py:58]: Ja volgens mij wel. Wat zij deden door layer -3 als uitgangspunt te nemen, betekent eigenlijk dat je de scores voor de oorspronkelijke 2622 identities als de features van je embedding gaat nemen.

File: lr_face/utils.py [hunk 8]
```diff
@@ -185,3 +185,12 @@ def resize_and_normalize(img, target_size):
         img_pixels /= 255
 
         return img_pixels
+
+
+def fix_tensorflow_rtx():
```
Comment (rolfypma) [inline lr_face/utils.py:190]: ik neem deze code voor waar aan

File: tests/src/util.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    :param relative_path: str
+    :return: str
+    """
+    tests_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    return os.path.join(tests_path, relative_path)
+
+
+def scratch_dir(path: str) -> Generator[str, None, None]:
+    """
```
Comment (rolfypma) [inline tests/src/util.py:19]: nice

File: lr_face/losses.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        Keras implementation of the triplet loss.
+
+        See https://en.wikipedia.org/wiki/Triplet_loss for more info.
+
+        Arguments:
+            y_true: Ignored, but required due to `Loss.call()` interface.
+            y_pred: A 3D tensor of shape `(batch_size, 3, embedding_size)`.
+                It contains the embeddings for each combination of anchor,
+                positive and negative image in the batch, respectively, which
+                is represented by the last axis of size 3. Make sure that each
```
Comment (rolfypma) [inline lr_face/losses.py:34]: je voert hier de l2 normalisatie niet uit, en checkt er ook niet op? zou dat niet moeten? Waar doen we het wel?

Comment (JeroenMandersloot) [reply lr_face/losses.py:34]: Erop checken in de loss vind ik wel een goed idee. Het uitvoeren ervan vind ik een verantwoordelijkheid van het model (bij `VGGFace` gebeurt dit nu al).

File: lr_face/utils.py (hunk 8 above)
Comment (JeroenMandersloot) [reply lr_face/utils.py:190]: Ja dit is een stukje code dat ik in elk project kopieer omdat ik een RTX kaart heb.

File: lr_face/models.py (hunk 7 above)
Comment (JeroenMandersloot) [reply lr_face/models.py:66]: Ik kan het commentaar over de inputs inderdaad beter hier weghalen, deze docstring op class-niveau moet meer een high-level uitleg zijn over wat deze class doet zonder al te veel op de implementatie van de individuele methodes in te gaan.

File: lr_face/models.py (hunk 6 above)
Comment (JeroenMandersloot) [reply lr_face/models.py:63]: Goed idee! Dan wordt uit de naam ook duidelijk dat het specifiek voor triplets werkt en eigenlijk niet voor andere manieren van finetunen.

File: finetuning.py (hunk 5 above)
Comment (JeroenMandersloot) [reply finetuning.py:37]: Hmm goed idee. Ga ik naar kijken.

- [ ] case-insensitive enum members

File: lr_face/losses.py (hunk 10 above)
Comment (JeroenMandersloot) [reply lr_face/losses.py:34]: Misschien is het beter om hier aparte unit tests voor te schrijven ipv in de triplet loss te testen. In principe is het namelijk niet iets wat de triplet loss per se zou moeten afdwingen.

Comment (rolfypma) [reply lr_face/losses.py:34]: nee maar omdat je specifiek de euclidische afstand neemt ben je er wel gevoelig voor. Zou het niet raar vinden iig een warning te krijgen

Comment (JeroenMandersloot) [reply lr_face/losses.py:34]: Ik heb het nu zo gemaakt dat je aan de `__init__()` van `TripletLoss` een boolean mee kunt geven of je wil afdwingen dat de input genormaliseerd is. Deze staat standaard op `True` (dus er wordt standaard wel gecontroleerd). Als blijkt dat de input niet genormaliseerd is, wordt er een exception geraised.
